[JENKINS-29547] plugins can't set an assigned node of 'master'

mdonohue opened the pull request:
Plugins need to use the setter/getter pair, because the 'assignedNode'
field is final.  But the AbstractProject uses different functions to read
and write the value, and didn't previously follow the same invariants as
the get/set pair.
getAssignedLabelString() for example, treats a true 'canRoam' the same as
a null 'assignedNode' field, but these represent different things.
This patch fixes getAssignedLabelString, so that users viewing the config
page will see the right string.

mdonohue committed:
[JENKINS-29547] plugins can't set an assigned node of 'master'

Plugins need to use the setter/getter pair, because the 'assignedNode'
field is final.  But the AbstractProject uses different functions to read
and write the value, and didn't previously follow the same invariants as
the get/set pair.
getAssignedLabelString() for example, treats a true 'canRoam' the same as
a null 'assignedNode' field, but these represent different things.
This patch fixes getAssignedLabelString, so that users viewing the config
page will see the right string.

oleg-nenashev commented:
No, it won't work. If I understand the change correctly, projects without label expression will be running on master only instead of "any node".

mdonohue committed:
Update getAssignedLabelString test - 'master' is no longer special case

These test cases were submitted in a large batch in 2013, so it's not
clear that they were testing anything other than the code "as written"
from before.  I am not seeing developer intent here, so this updates the
test to match the new logic that works better on the config page.

mdonohue commented:
Compare AbstractProject.getAssignedLabel() and AbstractProject.getAssignedLabelString().   This is making getAssignedNodeString correctly reflect in the configuration UI what is actually happening in getAssignedNode.  'getAssignedNode' is not changed here at all.

oleg-nenashev commented:
I'm aware about possible regressions due to the misusage of the method. I'll try to investigate it later.

Additional reviews will be appreciated.

mdonohue commented:
What kind of method misuse are you talking about?  I'd like to keep this moving, so if I can help with the research, let me know.

jtnord commented on test/src/test/java/hudson/model/ProjectTest.java:
```diff
@@ -175,9 +175,9 @@ public void testGetAssignedLabel() throws Exception{
         FreeStyleProject p = j.createFreeStyleProject("project");
         Slave slave = j.createOnlineSlave();
-        assertNull("Project should not have any label.", p.getAssignedLabelString());
+        assertEquals("Project should have jenkins master label.", j.jenkins.getSelfLabel().name, p.getAssignedLabelString());
```
at first glance this appears wrong - so the message should be clearer.
A project with no node assigned and a Jenkins with **no slaves** should have the jenkins master label.
A project with no node assigned and a Jenkins with nodes should not have a label assigned - as canRoam should be `true`

OTL Now looking at this class there seems to be like a bug in that when you add slaves and set the master to zero executors - the jobs will no longer build and you have to re-save every single one of them - but that is out of scope here.

mdonohue replied:
jtnord, I'm not familiar with the behavior described here - where a project with no node assigned has different tied-label behavior depending on whether Jenkins has slaves or not. 

I just looked at uses of the canRoam field - it is both private and volatile.    the 'setAssignedLabel' method does not take the slave count into consideration when setting canRoam to true or false.    It just checks whether the provided label argument is null or non-null.

jtnord commented:
I think this looks ok - apart from the unit test which needs to be a little clearer why it is expecting a job with no label restriction to be restricted to the master.

SO the question is should this method match `getAssignedLabel` or should `getAssingedLabel` be updated to do the same thing as `getAssingedLabelString`

jtnord commented on test/src/test/java/hudson/model/ProjectTest.java:
```diff
@@ -175,9 +175,9 @@ public void testGetAssignedLabel() throws Exception{
+        assertEquals("Project should have jenkins master label.", j.jenkins.getSelfLabel().name, p.getAssignedLabelString());
         p.setAssignedLabel(j.jenkins.getSelfLabel());
-        assertNull("Project should return null, because assigned label is Jenkins.", p.getAssignedLabelString());
+        assertEquals("Project should have jenkins master label after set explicitly", j.jenkins.getSelfLabel().name, p.getAssignedLabelString());
```
there is a missing test case here - if the project is saved again it should not have any assigned label.

mdonohue replied:
I'm not following why a save event would change the assigned label.

mdonohue commented:
It is not clear what is needed to get this accepted.  It is clear to me the current behavior in Jenkins core is broken - no plugin can set the slave node to 'master'.   This isn't a problem within core, becaues core doesn't use the public accessors to change slave nodes, while plugins are forced to use the public accessors.    The patch has languished for nearly a year now.  I'm throwing up my hands.

oleg-nenashev commented:
I'm not convinced regarding the fix feasibility. Maybe @jtnord or other @jenkinsci/code-reviewers have different opinion. Generally it still needs votes

daniel-beck commented:
@oleg-nenashev Why did you remove `needs-testcase`? It's not obvious this retains expected behavior while fixing the reported issue.

oleg-nenashev commented:
> OT: Now looking at this class there seems to be like a bug in that when you add slaves and set the master to zero executors - the jobs will no longer build and you have to re-save every single one of them - but that is out of scope here.

It is not longer allowed in Jenkins. See discussions in https://github.com/jenkinsci/jenkins/pull/3141 and https://github.com/jenkinsci/jenkins/pull/2925 . 

I do not really understand my comments above tho

daniel-beck commented:
Closing this given the age of the latest change and because there was no consensus how, or interest in moving this forward. @mdonohue I'm sorry we couldn't get this moved forward in a timely manner.

I still like this fix but it looks like this requires some more analysis to get confidence into what behavior is expect here. IMO the core API sucks worse than it needs to.
